### ISSUE TITLE
Error handling in tokenization

### ISSUE THREAD

**Issue by monzug** [alpheios-project/alignment-editor-new#140]
It would be nice if the tokenization process would generate some errors in the log, not just in the console log.
1) if I upload a tsv file as json file, nothing happens and I do not see any tokenization error. if I look in the web console, there is an error generated. I know that a tsv file should be uploaded as tsv and not json, but it can happen the json option is already checked because of previous upload action.
In case I upload a json file as tsv, I get the following error: Uploaded file has wrong format for the type - plainSourceUploadFromFile.

see screenshot of error console

![Screen Shot 2021-02-01 at 2 01 49 PM](https://user-images.githubusercontent.com/41396793/106462972-10d27780-6497-11eb-9c0b-b8f39526113c.png)

2) in issue #94, there was a tokenization failure due to connection problems (?). there were few errors in the log but no errors in the alignment editor. I killed the process because it was taking too long. Wouldn't be nice to have an error in the alignment editor? not sure how easy it would be to implement this one, so 2) is a low priority, but 1) should not be.

**Comment by irina060981** [comment]
>if I upload a tsv file as json file, nothing happens and I do not see any tokenization error. if I look in the web console, there is an error generated. I know that a tsv file should be uploaded as tsv and not json, but it can happen the json option is already checked because of previous upload action.
In case I upload a json file as tsv, I get the following error: Uploaded file has wrong format for the type - plainSourceUploadFromFile.

Fixed with changing the upload workflow - https://github.com/alpheios-project/alignment-editor-new/issues/135#issuecomment-771560267

**Comment by irina060981** [comment]
>in issue #94, there was a tokenization failure due to connection problems (?). there were few errors in the log but no errors in the alignment editor. I killed the process because it was taking too long. Wouldn't be nice to have an error in the alignment editor? not sure how easy it would be to implement this one

@monzug, I am not sure how to reproduce it - because we should get some response with timeout error from the remote source or connection error.
If we get one of them - it would be published to the editor.

Otherwise - I should paste some timeout error (and define time period for that) on our side - I am not sure that it is a good practice.

I suggest to set only a limit for the text length for now, as you and @balmas suggested here - https://github.com/alpheios-project/alignment-editor-new/issues/133#issuecomment-770833844

**Comment by irina060981** [comment]
About setting the limit - 
I used tests from @abrasax
(attached to the issues)

Greek has 34729 characters (with spaces)
Chineese has 10787 characters (with spaces)

Greek needs near 15-23 seconds for tokenization
Chinese needs near 9-20 seconds for tokenization

If upload it and tokenize inside editor - it would work slowly for me (even getting right mouse menu button) - because of a big amount of tokens - but it works.

So the main problem here is the amount of tokens (in sum of all texts).
I was to wait even the right button clicked menu to get Inspector and console logs.

Chrome got near a 1GB for this page.

So we should decide what amount of characters should be set as a limit for a single text - if we won't to set such a limit.

[euthyphroChinese.txt](https://github.com/alpheios-project/alignment-editor-new/files/5910905/euthyphroChinese.txt)
[euthyphroGreek.txt](https://github.com/alpheios-project/alignment-editor-new/files/5910906/euthyphroGreek.txt)

**Comment by monzug** [comment]
I sent an email too.
"I uploaded a txt file and then downloaded as json and uploaded again with the alignment groups.
The word count was 5,293 words and 32,127 characters with spaces. It worked and tokenization was fast. However I got the message:
This webpage is using significant memory. Closing it may improve the responsiveness of your MAC.
see attachment for word count"
![Screen Shot 2021-02-03 at 3 18 19 PM](https://user-images.githubusercontent.com/41396793/106763002-af490f00-6636-11eb-82ea-c3c80d3c7f0a.png)

**Comment by balmas** [comment]
per discussion with @abrasax today, as @irina060981 previously suggested, we'd like to set a default limit of X number of words, making this configurable by the user.  We will defer to @monzug on what a good starting limit should be (I think maybe we just need one limit, for the entire alignment regardless of the number of texts?)

**Comment by monzug** [comment]
maybe trying to upload different size, till I do not get the above message. what do you think?

**Comment by balmas** [comment]
yes, thank makes sense. Maybe also trying with different numbers of texts (up to e.g 4 translations?)

**Comment by monzug** [comment]
I remember doing a test with 4 or 5 translation and it worked (I think I
sent you an email too). anyhow, good idea.

**Comment by abrasax** [comment]
Would the limit apply to the total number of tokens across all translations
or simply to each individual translation? I do hope the latter.

**Comment by balmas** [comment]
Well, I think the limit is going to have to apply to the total number of tokens -- you can have a lot of small texts or fewer larger texts.

I think there are probably some optimizations @irina060981 can do to improve things but I would recommend focusing on the display functionality first, and then returning to the performance question after the major required demo features are in place.  The point of a limit right now is just to keep the tool running smoothly while demoing.

**Comment by abrasax** [comment]
agreed. I have 20 small translations I should like to be able to show.
I assume for languages with spaces between the words there is no need to
specify which language being aligned?

**Comment by balmas** [comment]
Well if you want to get the language-specific tokenization (if the space
libraries support that language) you should. Otherwise you will get English
rules applied.

**Comment by irina060981** [comment]
>I think there are probably some optimizations @irina060981 can do to improve things

From my tests the main problems are with rendering - to much floating (display: inline-block) tags, browser would need to do much calculations. I am not sure but I would do some tests may be some changes to layout could do make it easier. But anyway text rendering is the most dificcult for browsers (no limits and a need to check scrolls).

As for other kind of optimization:
- I could do rendering tokens editor by request (I tried, but there were some problems - I reverted and decided to finih with output HTML first). Also From my tests it didn't help much
- I could do loading by segments, sentences - but I believe, that it is a common way for optimization but is not liked much by users :)

**Comment by irina060981** [comment]
Described optimization are made

**Comment by monzug** [comment]
Irina, which type of optimization did you do? how can I test it? Thanks

**Comment by irina060981** [comment]
I did the following:
- I made some changes to technical layout - you won't be able to check it
- tokens editor is rendered only after show is clicked once - may be you could check that this works properly?

**Comment by monzug** [comment]
I like that we need to click on Show now. will do more testing on this.

**Comment by monzug** [comment]
this works in all browser. tested in  Alpheios Translation Alignment editor 1.2.2 (build development.20210311357)
